File: Embedded/Networking/src/com/projectgame/networking/IXPFile.java
package com.projectgame.networking;

import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.*;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.*;
import java.util.Hashtable;

/**
 * Created by Beppo-Laptop on 10/8/2015.
 */
public class IXPFile {
    private static final String XML_HEADER = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>";
    private static final String XML_IXP = "IXP";
    private static final String XML_IXP_TARGET = "Target";
    private static final String XML_IXP_TARGET_NFUNCTION = "nfunction";
    private static final String XML_IXP_HEADER = "Header";
    private static final String XML_IXP_HEADER_INFO = "Info";
    private static final String XML_IXP_HEADER_INFO_NAME = "name";
    private static final String XML_IXP_HEADER_INFO_VALUE = "value";
    private static final String XML_IXP_BODY = "Body";
    private static final String XML_IXP_BODY_INFO = "Info";
    private static final String XML_IXP_BODY_INFO_NAME = "name";

    private static final String XML_BASE =
            XML_HEADER +
            "<" + XML_IXP + ">" +
                "<"  + XML_IXP_TARGET + " " + XML_IXP_TARGET_NFUNCTION + "=\"\"" + "/>" +
                "<"  + XML_IXP_HEADER + ">" +
                "</" + XML_IXP_HEADER + ">" +
                "<"  + XML_IXP_BODY + ">" +
                "</" + XML_IXP_BODY + ">" +
            "</" + XML_IXP + ">";

    public static IXPFile createNew() throws IOException, SAXException, ParserConfigurationException {
        return parse(XML_BASE);
    }
    public static IXPFile parse(String source) throws ParserConfigurationException, IOException, SAXException {
        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        DocumentBuilder db = dbf.newDocumentBuilder();

        byte[] data = source.getBytes("UTF-8");
        ByteArrayInputStream bin = new ByteArrayInputStream(data);

        Document doc = db.parse(bin);

        return new IXPFile(doc);
    }

    private Document doc;

    private IXPFile(Document doc) {
        this.doc = doc;
    }

    public String getXML() throws TransformerException {
        TransformerFactory tf = TransformerFactory.newInstance();
        Transformer transformer = tf.newTransformer();
        transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
        StringWriter writer = new StringWriter();
        transformer.transform(new DOMSource(doc), new StreamResult(writer));
        String output = writer.getBuffer().toString().replace("\n|\r", "");

        return XML_HEADER + output;
    }

    public String getNetworkFunction(){
        return getNFunctionAttr().getNodeValue();
    }
    public void setNetworkFunction(String networkFunction){
        getNFunctionAttr().setNodeValue(networkFunction);
    }

    public Hashtable<String, String> getHeaders(){
        Node header = getHeaderNode();
        NodeList childs = header.getChildNodes();
        Hashtable<String, String> headers = new Hashtable<>();

        for(int i = 0; i < childs.getLength(); i++){
            Node info = childs.item(i);
            Node name = info.getAttributes().getNamedItem(XML_IXP_HEADER_INFO_NAME);
            Node value = info.getAttributes().getNamedItem(XML_IXP_HEADER_INFO_VALUE);

            headers.put(name.getNodeValue(), value.getNodeValue());
        }
        return headers;
    }
    public Hashtable<String, String> getInfos(){
        Node body = getBodyNode();
        NodeList list = body.getChildNodes();
        Hashtable<String, String> infos = new Hashtable<>();

        for(int i = 0; i < list.getLength(); i++){
            Node info = list.item(i);
            Node name = info.getAttributes().getNamedItem(XML_IXP_BODY_INFO_NAME);
            Node value = info.getFirstChild();

            infos.put(name.getNodeValue(), value.getNodeValue());
        }

        return infos;
    }

    public void addHeader(String key, String value){
        Node header = getHeaderNode();

        Node nameAttr = doc.createAttribute(XML_IXP_HEADER_INFO_NAME);
        Node valueAttr = doc.createAttribute(XML_IXP_HEADER_INFO_VALUE);

        nameAttr.setNodeValue(key);
        valueAttr.setNodeValue(value);

        Node newInfo = doc.createElement(XML_IXP_HEADER_INFO);

        newInfo.getAttributes().setNamedItem(nameAttr);
        newInfo.getAttributes().setNamedItem(valueAttr);

        header.appendChild(newInfo);
    }
    public void addInfo(String key, String value){
        Node header = getBodyNode();

        Node nameAttr = doc.createAttribute(XML_IXP_HEADER_INFO_NAME);
        nameAttr.setNodeValue(key);

        Node newInfo = doc.createElement(XML_IXP_HEADER_INFO);
        newInfo.getAttributes().setNamedItem(nameAttr);;

        Node val = doc.createTextNode(value);
        newInfo.appendChild(val);

        header.appendChild(newInfo);
    }

    private Node getRootNode(){
        return doc.getDocumentElement();
    }
    private Node getTargetNode(){
        NodeList list = getRootNode().getChildNodes();

        for(int i = 0; i < list.getLength(); i++){
            Node node = list.item(i);

            if(node.getNodeName().equals(XML_IXP_TARGET))
                return node;
        }

        return null;
    }
    private Node getNFunctionAttr(){
        Node node = getTargetNode();
        NamedNodeMap attributes = node.getAttributes();
        Node attribute = attributes.getNamedItem(XML_IXP_TARGET_NFUNCTION);

        return attribute;
    }
    private Node getHeaderNode(){
        NodeList list = getRootNode().getChildNodes();

        for(int i = 0; i < list.getLength(); i++){
            Node node = list.item(i);

            if(node.getNodeName().equals(XML_IXP_HEADER))
                return node;
        }

        return null;
    }
    private Node getBodyNode(){
        NodeList list = getRootNode().getChildNodes();

        for(int i = 0; i < list.getLength(); i++){
            Node node = list.item(i);

            if(node.getNodeName().equals(XML_IXP_BODY))
                return node;
        }

        return null;
    }
}

File: ES_Py/Core/ESApi/IXPFile.py
from xml.dom import minidom

class IXPFile:
    __network_function = ""  # type: str
    __header_names = []  # type: List[str]
    __header_values = []  # type: List[str]
    __info_names = []  # type: List[str]
    __info_values = []  # type: List[str]

    __XML_HEADER = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
    __XML_IXP = "IXP"
    __XML_IXP_TARGET = "Target"
    __XML_IXP_TARGET_NFUNCTION = "nfunction"
    __XML_IXP_HEADER = "Header"
    __XML_IXP_HEADER_INFO = "Info"
    __XML_IXP_HEADER_INFO_NAME = "name"
    __XML_IXP_HEADER_INFO_VALUE = "value"
    __XML_IXP_BODY = "Body"
    __XML_IXP_BODY_INFO = "Info"
    __XML_IXP_BODY_INFO_NAME = "name"

    __XML_BASE = \
            __XML_HEADER + \
            "<" + __XML_IXP + ">" + \
                "<"  + __XML_IXP_TARGET + " " + __XML_IXP_TARGET_NFUNCTION + "=\"\"" + "/>" + \
                "<"  + __XML_IXP_HEADER + ">" + \
                "</" + __XML_IXP_HEADER + ">" + \
                "<"  + __XML_IXP_BODY + ">" + \
                "</" + __XML_IXP_BODY + ">" + \
            "</" + __XML_IXP + ">"

    def __init__(self, source = None):
        if source is None:
            return

        doc = minidom.parseString(source)

        nFuncNode = doc.getElementsByTagName(self.__XML_IXP_TARGET)[0]
        self.__network_function = nFuncNode.getAttribute(self.__XML_IXP_TARGET_NFUNCTION)

        headerNode = doc.getElementsByTagName(self.__XML_IXP_HEADER)[0]
        headerNodes = headerNode.getElementsByTagName(self.__XML_IXP_HEADER_INFO)

        for curNode in headerNodes:
            name = curNode.getAttribute(self.__XML_IXP_HEADER_INFO_NAME)
            value = curNode.getAttribute(self.__XML_IXP_HEADER_INFO_VALUE)
            self.__header_names.append(name)
            self.__header_values.append(value)

        bodyNode = doc.getElementsByTagName(self.__XML_IXP_BODY)[0]
        bodyNodes = bodyNode.getElementsByTagName(self.__XML_IXP_BODY_INFO)


        for curNode in bodyNodes:
            name = curNode.getAttribute(self.__XML_IXP_BODY_INFO_NAME)
            value = curNode.childNodes[0].nodeValue
            self.__info_names.append(name)
            self.__info_values.append(value)

    def get_xml(self) -> str:
        doc = minidom.parseString(self.__XML_BASE)

        nFuncNode = doc.getElementsByTagName(self.__XML_IXP_TARGET)[0]
        self.__network_function = nFuncNode.setAttribute(self.__XML_IXP_TARGET_NFUNCTION, self.__network_function)

        for index in range(0, self.headers_count()):
            headerNode = doc.createElement(self.__XML_IXP_HEADER_INFO)
            headerNode.setAttribute(self.__XML_IXP_HEADER_INFO_NAME, self.header_name_at(index))
            headerNode.setAttribute(self.__XML_IXP_HEADER_INFO_VALUE, self.header_value_at(index))

            doc.getElementsByTagName(self.__XML_IXP_HEADER)[0].appendChild(headerNode)

        for index in range(0, self.infos_count()):
            infoNode = doc.createElement(self.__XML_IXP_BODY_INFO)
            infoNode.setAttribute(self.__XML_IXP_BODY_INFO_NAME, self.info_name_at(index))

            infoTextNode = doc.createTextNode(self.info_value_at(index))
            infoNode.appendChild(infoTextNode)

            doc.getElementsByTagName(self.__XML_IXP_BODY)[0].appendChild(infoNode)

        return doc.toxml()

    def get_network_function(self) -> str:
        return self.__network_function

    def headers_count(self) -> int:
        return len(self.__header_names)

    def header_name_at(self, index: int) -> str:
        if index < 0 or index >= self.headers_count():
            raise IndexError()

        return self.__header_names[index]

    def header_value_at(self, index: int) -> str:
        if index < 0 or index >= self.headers_count():
            raise IndexError()

        return self.__header_values[index]

    def infos_count(self) -> int:
        return len(self.__info_names)

    def info_name_at(self, index: int) -> str:
        if index <  0 or index >= self.infos_count():
            raise IndexError()

        return self.__info_names[index]

    def info_value_at(self, index: int) -> str:
        if index < 0 or index >= self.infos_count():
            raise IndexError()

        return self.__info_values[index]

    def add_info(self, name: str, value: str):
        self.__info_names.append(name)
        self.__info_values.append(value)

    def add_header(self, name: str, value: str):
        self.__header_names.append(name)
        self.__header_values.append(value)